### PR TITLE
Implement changes for v2

### PR DESCRIPTION
**Pull request by JensSkipr**
- Converters: Add `ToSnakeCase`
- Logging: Add `AddLoggingInfo*` helpers
- Metadata: Add package to support uniform metadata
- Align everything with new `metadata` package

Main change is the new `metadata` package. This allows us to have a uniform way of working with metadata, independent of the used packages (Gin, go-micro, ...). It's build as modular as possible which allows easy converters for metadata between packages. E.g.
```go
func ConvertGinToGoMicro(c *gin.Context) context.Context {
	// Get base context from request
	ctx := c.Request.Context()

	// "genErr" ignored since it can only fail on extracting metadata from context.
	// This cannot occur since there is no go-micro metadata on the context yet.
	meta := GetGinMetadata(c)
	ctx, _, _ = UpdateGoMicroMetadata(ctx, meta)

	// Return result
	return ctx
}
```

### CODE REVIEW

**Comment by JensSkipr** [comment]
Fixes https://github.com/skiprco/go-utils/issues/18

File: converters/strings.go [hunk 1]
```diff
@@ -25,3 +27,15 @@ func NormaliseString(input string) (string, *errors.GenericError) {
 	// Normalise successful
 	return output, nil
 }
+
+var matchFirstCap = regexp.MustCompile("([A-Z])([A-Z][a-z])")
+var matchAllCap = regexp.MustCompile("([a-z0-9])([A-Z])")
+
+// ToSnakeCase converts the provided string to snake_case.
+// Based on https://gist.github.com/stoewer/fbe273b711e6a06315d19552dd4d33e6
+func ToSnakeCase(input string) string {
```
Comment (ludoviclaffineur) [inline converters/strings.go:36]: nice function

File: logging/audit_helpers.go [hunk 2]
```diff
@@ -94,22 +94,31 @@ func logEvent(ctx context.Context, message string, category AuditCategory, addit
...
-		maxLength := len(meta) + len(additionalData) + 2 // for message and category
+	meta, genErr := metadata.GetGoMicroMetadata(ctx)
+	if genErr != nil {
+		// Frontend should not handle errors regarding audit logging.
+		// Therefore, we log an error and ignore the metadata from the context.
+		log.WithField("error", genErr).Error("Failed to extract metadata from go-micro context for audit logging")
+		meta = metadata.Metadata{}
+	}
+
+	if len(meta) != 0 {
```
Comment (ludoviclaffineur) [inline logging/audit_helpers.go:105]: is this check really needed I think if the length is 0 the for would be skipped?

```golang

maxLength := len(meta) + len(additionalData) + 1 // for category
logFields = make(log.Fields, maxLength)
for k, v := range meta {
	snakeKey := converters.ToSnakeCase(k)
	logFields[snakeKey] = v
}
```

**Review by ludoviclaffineur — COMMENTED** [review]
Small remark/question but overall nice pr.

File: logging/audit_helpers.go (hunk 2 above)
Comment (JensSkipr) [reply logging/audit_helpers.go:105]: Nice catch! You're correct, `range meta` will automatically be skipped.

Comment (JensSkipr) [reply logging/audit_helpers.go]: Fixed by b3914df

File: converters/strings.go (hunk 1 above)
Comment (JensSkipr) [reply converters/strings.go:36]: A bit larger that the promised 2 lines, but I prefer readability in this case as a single line would be really ugly :smile:

**Pull request merged**